### PR TITLE
[DTensor] Propagate 'grad_dtype' attribute in to_local/from_local/redistribute

### PR DESCRIPTION
**Pull request by SongyuanZhao**
This PR enables the preservation of gradient precision (preventing downcasting) during DTensor conversions by explicitly handling the grad_dtype attribute.(#173990)

**Key Changes:**

1.Removed grad_dtype validation for non-leaf nodes to support change 2
2.Propagated grad_dtype in DTensor operations: Updated to_local, from_local, and redistribute methods to pass and set the grad_dtype attribute correctly. 

**Example:**
Here is an example demonstrating this in an FSDP + EP scenario:
**AS IS:**
```text
W(dtensor) ---> (to_local) ---> W(tensor) -------> (gmm) --------> Y
dtype=bf16                      dtype=bf16                    dtype=bf16
grad_dtype=fp32

(leaf node)                     (non-leaf)                       
                                                                   
dW(dtensor) <--(to_local_bwd)-- dW(tensor) <------(gmmbwd) <------ dY
dtype=fp32                      dtype=bf16      with fp32 out   dtype=bf16
      ^                               ^                         
      |____[upcast to fp32]          |____[downcast to bf16] (align to W dtype)
```
**TO BE:**
Since grad_dtype is passed through to_local, backward gradients are no longer downcasted
```text
W(dtensor) ---> (to_local) ---> W(tensor) -------> (gmm) --------> Y
dtype=bf16                      dtype=bf16                    dtype=bf16
grad_dtype=fp32                 grad_dtype=fp32

(leaf node)                     (non-leaf)                       
                                                                   
dW(dtensor) <--(to_local_bwd)-- dW(tensor) <------(gmmbwd) <------ dY
dtype=fp32                      dtype=fp32      with fp32 out   dtype=bf16
                                      ^                         
                                      |____[no cast] (align to W grad_dtype)
```

### CODE REVIEW

**Review by wconstab — COMMENTED** [review]
this looks ok to me! 
@SongyuanZhao can you add a PR description saying what the change does

I'm gonna see if @soulitzer or @albanD can review just in case there is a better way to accomplish this since i'm not an autograd expert

File: test/test_autograd.py [hunk 1]
```diff
@@ -3834,6 +3833,29 @@ def backward(ctx, grad_output):
         z2.sum().backward()
         self.assertEqual(l.grad.dtype, torch.float32)
 
+    @skipIfTorchDynamo("grad_dtype not supported in compile")
```
Comment (soulitzer) [inline test/test_autograd.py]: This test is probably redundant given that we have modified the previous test

File: torch/csrc/autograd/python_variable.cpp [hunk 2]
```diff
@@ -3070,8 +3070,6 @@ static PyObject* THPVariable_get_grad_dtype(THPVariable* self, void* unused) {
     return handle_torch_function_getter(self, "grad_dtype");
   }
   const auto& var = THPVariable_Unpack(self);
-  TORCH_CHECK(
-      !var.grad_fn(), "grad_dtype can only be accessed on leaf tensors.");
```
Comment (soulitzer) [inline torch/csrc/autograd/python_variable.cpp:3074]: When would you want to check grad-dtype on a non-leaf?

Comment (soulitzer) [reply torch/csrc/autograd/python_variable.cpp:3074]: The reason we support grad-dtype for leaf tensors is for example, local microbatching, where we want to accumulate in a higher-precision  into the .grad field. But there is no .grad field on non-leaf (unless you retain_grad()).

File: torch/distributed/tensor/_api.py [hunk 3]
```diff
@@ -226,6 +228,7 @@ def forward(  # type: ignore[override]
             # pyrefly: ignore [unexpected-keyword]
             requires_grad=input.requires_grad,
         )
+        dist_tensor.grad_dtype = input.grad_dtype
```
Comment (soulitzer) [inline torch/distributed/tensor/_api.py:231]: hmm this is actually secretly circumventing the check today that this attribute can only be set on leaf tensors

File: torch/csrc/autograd/python_variable.cpp (hunk 2 above)
Comment (sorenzzio) [reply torch/csrc/autograd/python_variable.cpp:3074]: The main goal here is to ensure that the gradients output high precision until they are accumulated into .grad. For example, my mmbwd naturally supports outputting FP32 gradients, and I don't want them cast to BF16 before accumulation, as that leads to precision loss. 
However, in FSDP scenarios, the insertion of dtensor's to_local/from_local operations between the parameters and the computation inadvertently causes this downcasting to occur.

**Comment by SongyuanZhao** [comment]
I've updated the description to include an example that better explains this PR.

File: torch/distributed/tensor/_api.py (hunk 3 above)
Comment (sorenzzio) [reply torch/distributed/tensor/_api.py:231]: This is indeed a bit tricky, grad_dtype is set before grad_fn is attached.

File: test/test_autograd.py (hunk 1 above)
Comment (sorenzzio) [reply test/test_autograd.py]: Removed
